bigint/big_integer.cpp

```diff
@@ -2,241 +2,429 @@
+    big_integer real_a = additional(a.size < b.size ? b : a);
+    big_integer real_b = additional(a.size < b.size ? a : b);
+    if (real_a.size < real_b.size) {
+        swap(real_a, real_b);
+    }
+    big_integer answer;
+    for (size_t i = 0; i < real_b.size; i++) {
+        answer.push_back(function(real_a[i], real_b[i]));
+    }
+    uint32_t help = real_b.sign ? UINT32_MAX : 0;
+    for (size_t i = real_b.size; i < real_a.size; i++) {
+        answer.push_back(function(real_a[i], help));
+    }
+    answer.sign = function(a.sign, b.sign);
+    answer.normalise();
+    return additional(answer);
 }
 
-big_integer operator^(big_integer a, big_integer const& b)
-{
-    return a ^= b;
+big_integer operator&(big_integer const &a, big_integer const &b) {
+    return bit_operation(a, b, [](uint32_t first, uint32_t second) { return first & second; });
 }
 
-big_integer operator<<(big_integer a, int b)
-{
-    return a <<= b;
+big_integer operator|(big_integer const &a, big_integer const &b) {
+    return bit_operation(a, b, [](uint32_t first, uint32_t second) { return first | second; });
 }
 
-big_integer operator>>(big_integer a, int b)
-{
-    return a >>= b;
+big_integer operator^(big_integer const &a, big_integer const &b) {
+    return bit_operation(a, b, [](uint32_t first, uint32_t second) { return first ^ second; });
 }
 
-bool operator==(big_integer const& a, big_integer const& b)
-{
-    return mpz_cmp(a.mpz, b.mpz) == 0;
+big_integer operator<<(big_integer const &a, int b) {
+    if (b > 0) {
+        big_integer answer = a;
+        uint32_t small = b % 32, big = b / 32, short_multiply = (1u << small);
+        answer = increase(answer, short_multiply);
+        answer.reverse();
+        for (size_t i = 0; i < big; i++) {
+            answer.push_back(0);
+        }
+        answer.reverse();
+        answer.sign = a.sign;
+        return answer;
+    } else {
+        return a >> -b;
+    }
 }
 
-bool operator!=(big_integer const& a, big_integer const& b)
-{
-    return mpz_cmp(a.mpz, b.mpz) != 0;
+big_integer operator>>(big_integer const &a, int b) {
+    if (b > 0) {
+        big_integer answer = a;
+        uint32_t small = b % 32, big = b / 32, short_divide = (1u << small), length = answer.size;
+        answer.reverse();
+        for (size_t i = 0; i < std::min(big, length); i++) {
+            answer.bits.pop_back();
+            answer.size--;
+        }
+        answer.reverse();
+        answer = short_div(answer, short_divide);
+        answer.sign = a.sign;
+        answer.normalise();
+        return answer;
+    } else {
+        return a << -b;
+    }
 }
 
-bool operator<(big_integer const& a, big_integer const& b)
-{
-    return mpz_cmp(a.mpz, b.mpz) < 0;
+
+bool operator==(big_integer const &a, big_integer const &b) {
+    if (a.size != b.size || a.sign != b.sign) {
+        return false;
+    }
+    for (size_t i = a.size; i > 0; i--) {
```

Comment:
[minor] [style]

```suggestion
    for (size_t i = a.size; i-- > 0;) {
```

и ниже `a[i]`